File: euler_go/src/euler/problem67.go
// problem67.go
//
// Find the maximum total from top to bottom in triangle.txt a 15K text
// file containing a triangle with one-hundred rows.

package main

import (
	"bufio"
	"euler/tools"
	"fmt"
	"log"
	"os"
	"strings"
)

// Triangle is a list of lists of ints such as [[1] [2 3] [4 5 6]]
// representing a triangle of the form:
//     1
//    2 3
//   4 5 6
type Triangle [][]int

var cache = make(map[string]int)

// Return a key suitable for use in a map.
// (Note: currently very slow.)
func k(tr Triangle) string {
	return fmt.Sprintf("%v", tr)
}

// Recursively find the maximum value of the root node plus the largest
// of its children, and so on, all the way to the base.
func maxRoute(tr Triangle) int {
	if val, ok := cache[k(tr)]; ok {
		return val
	}
	root := tr[0][0]
	if len(tr) == 1 {
		return root
	}
	a, b := children(tr)
	result := root + tools.Max(maxRoute(a), maxRoute(b))
	cache[k(tr)] = result
	return result
}

// Split the triangle in two below the root node.
// e.g. [[1] [2 3] [4 5 6]] -> [[2] [4 5]] [[3] [5 6]]
func children(tr Triangle) (Triangle, Triangle) {
	var a, b Triangle
	for _, row := range tr[1:] {
		a = append(a, row[:len(row)-1])
		b = append(b, row[1:])
	}
	return a, b
}

func problem67() int {
	file, err := os.Open("data/triangle.txt")
	if err != nil {
		log.Fatal(err)
	}
	defer file.Close()

	var tr Triangle
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		row := scanner.Text()
		ss := strings.Split(row, " ")
		tr = append(tr, tools.StringsToInts(ss))
	}
	if err := scanner.Err(); err != nil {
		log.Fatal(err)
	}
	return maxRoute(tr)
}

func main() {
	ans := problem67()
	fmt.Println(ans)
}

File: euler_python/problem67.py
"""
problem67.py

https://projecteuler.net/problem=67

Find the maximum total from top to bottom in triangle.txt a 15K text
file containing a triangle with one-hundred rows.
"""
from toolset import memoize_mutable

@memoize_mutable
def max_route(triangle):
    """Recursively find the maximum value of the root node plus the
    largest of its children, and so on, all the way to the base."""
    # where triangle is a list of lists such as [[1], [2, 3], [4, 5, 6]]
    # representing a tree of the form:
    #   1
    #  2 3
    # 4 5 6
    root = triangle[0][0]
    if len(triangle) == 1:
        return root
    a, b = children(triangle)
    return root + max(max_route(a), max_route(b))

def children(triangle):
    "Split the triangle in two below the root node"
    # [[1], [2, 3], [4, 5, 6]] --> [[2], [4, 5]], [[3], [5, 6]]
    # the two children triangles of the root node.
    a = [row[:-1] for row in triangle[1:]]
    b = [row[1:] for row in triangle[1:]]
    return a, b

def problem67():
    with open("data/triangle.txt", "r") as f:
        triangle = [list(map(int, row.split())) for row in f]
    return max_route(triangle)
